pkg/controller/migration/convert/felix_vars.go

```diff
@@ -0,0 +1,192 @@
+	return types.JSONPatchType
+}
+
+func (s *patches) Data(obj runtime.Object) ([]byte, error) {
+	return json.Marshal(s)
+}
+
+func handleFelixVars(c *components) error {
+	cn := getContainer(c.node.Spec.Template.Spec, containerCalicoNode)
+	if cn == nil {
+		return fmt.Errorf("missing calico-node container")
+	}
+	// loop through all env vars of the form 'FELIX_key=val', and convert them
+	// into patches
+	p := new(patches)
+	for _, env := range cn.Env {
+		if !strings.HasPrefix(env.Name, "FELIX_") {
+			continue
+		}
+
+		// skip any value that was otherwise checked / accounted for somewhere else
+		// in the migration code
+		if _, ok := c.node.checkedVars[containerCalicoNode].envVars[env.Name]; ok {
+			continue
+		}
+
+		fval, err := c.node.getEnv(ctx, c.client, containerCalicoNode, env.Name)
+		if err != nil {
+			return err
+		}
+
+		// downcase and remove FELIX_ prefix
+		key := strings.ToLower(strings.TrimPrefix(env.Name, "FELIX_"))
+		pp, err := patchFromVal(key, *fval)
+		if err != nil {
+			return err
+		}
+		*p = append(*p, pp)
+
+	}
+
+	return c.client.Patch(ctx, &crdv1.FelixConfiguration{
+		ObjectMeta: metav1.ObjectMeta{Name: "default"},
+	}, p)
+}
+
+func patchFromVal(key, val string) (patch, error) {
+	// since env vars are caps lock, we need to get the correct casing of
+	// the given env var. to do this, loop through the felixconfigspec
+	// using reflection, finding the struct field where the downcased name
+	// matches the downcased env var name.
+	fc := reflect.ValueOf(crdv1.FelixConfigurationSpec{})
+	for ii := 0; ii < fc.Type().NumField(); ii++ {
+		field := fc.Type().Field(ii)
+		value := fc.Field(ii)
+
+		if strings.ToLower(key) == strings.ToLower(field.Name) {
+			fieldName := strings.Split(field.Tag.Get("json"), ",")[0]
+
+			v, err := convert(value.Interface(), val)
+			if err != nil {
+				return patch{}, err
+			}
+
+			return patch{
+				Op:    "replace",
+				Path:  fmt.Sprintf("/spec/%s", fieldName),
+				Value: v,
+			}, nil
+		}
+	}
+
+	return patch{}, fmt.Errorf("unrecognized felix config setting: %v", key)
+}
+
+// convert transforms a string representation to the desired type <t>.
+func convert(t interface{}, str string) (interface{}, error) {
+	ptr := false
+	switch t.(type) {
+	case bool:
+		b, err := strconv.ParseBool(str)
+		if err != nil {
+			return nil, err
+		}
+		return b, nil
+	case *bool:
+		b, err := strconv.ParseBool(str)
+		if err != nil {
+			return nil, err
+		}
+		return &b, nil
+	case *uint32:
+		i, err := strconv.ParseInt(str, 10, 33)
+		if err != nil {
+			return nil, err
+		}
+		u := uint32(i)
+		return &u, nil
+
+	case int:
+		i, err := strconv.Atoi(str)
+		if err != nil {
+			return nil, err
+		}
+		return i, nil
+	case string:
+		if ptr {
+			return &str, nil
+		}
+		return str, nil
+	case *crdv1.IptablesBackend:
+		v := crdv1.IptablesBackend(str)
+		return &v, nil
+	case *crdv1.AWSSrcDstCheckOption:
+		v := crdv1.AWSSrcDstCheckOption(str)
+		return &v, nil
```

Comment:
I'm not sure if this is OK or if I need to run a switch statement on the value and return the actual 'enum' it corresponds to.